### ISSUE TITLE
Finetuning with Borzoi model and changing peak widths

### ISSUE THREAD

**Issue by kenxie7** [aertslab/CREsted#130]
### Report

Hi! I wanted to test for a different peak width (ie 500bp) with the peak regression model, however it seems that the dilated_cnn model only accepts input of 2114bp and from the tutorial, I understood that the model structure can be changed to fit the different widths. One of them seem to be the borzoi model (https://crested.readthedocs.io/en/stable/tutorials/borzoi_atac_finetuning.html), but as i was going through the tutorial, i couldn't find the model weights h5 file from the following code:
```
# Load pretrained Borzoi weights
model_file, _ = crested.get_model("Borzoi_mouse_rep0")
# Put weights into base architecture
with zipfile.ZipFile(
    model_file
) as model_archive, tempfile.TemporaryDirectory() as tmpdir:
    model_weights_path = model_archive.extract("model.weights.h5", tmpdir)
    base_model_architecture.load_weights(model_weights_path)
```

In the downloaded files, I only see .tsv file and a .keras file (borzoi_crested_mouse_fold0.keras and borzoi_mouse_output_classes.tsv). For a peak regression model with around 500bp, do you have other suggestions of model that might work in this case?

Many thanks!!

**Comment by LukasMahieu** [comment]
Hey,

There are multiple things going on here. I don't understand exactly what you're trying to achieve so I might be able to help better with some extra input, but here are some answers already:
1. All the model architectures in the `crested.tl.zoo` have a *seq_len* parameter that you can set to your region's width to change the input size before starting your training. The *dilated_cnn* model is a bit of a special case, since it uses a stack of dilated convolutions to process wider and wider region's width. 
By default the dilated_cnn architecture uses 8 dilation layers, but if you only have 500bp as inputs then by the sixth layer or so your model's receptive field has already grown larger than 500bp, causing an error, regardless what you set your input size to be. If you want to use dilated convolutions for 500bp input, you need to change both the *seq_len* and the *n_dil_layer* (e.g. set that to 4).
Using dilated convolutions for only 500bp inputs is probably a bit overkill. There are other architectures in the zoo that don't use them, such as the simple_convnet or deeptopic_cnn. Remember to set your output_activation parameter to "softplus" or something similar if you're training a regression model. The deeptopic_cnn for example has a default output_activation of "sigmoid', since historically we've used it more for classification tasks, but there is nothing inherent in the architecture that would not make it work for regression tasks (I've used it myself for regression models).
2. If your goal is to load one of our pretrained models from the repository and change the input size to test it on the same task it was trained on but on differently sized regions, then you're out of luck. There are workarounds though; the easiest would be to simply pad the input sequence with all "Ns" (which will be one hot encoded to 0s) or with randomly generated sequences to the size the model expects.
3. I'm not sure what you're trying to do with our pretrained borzoi model. All the models in the repository get downloaded as a zip and then unpacked to the model.keras file and the output_classes.tsv. You can simply load those with `keras.models.load_model(...)`. If you want only the weights then load it as a keras model and then access the layers' weights with model.layers.weights
Also, the default borzoi doesn't take 500bp as input but 524288bp [see here](https://crested.readthedocs.io/en/latest/models/EnformerBorzoi/borzoi.html). 
We have a finetuned version that uses 2048bp as inputs on the BICCN dataset [see here](https://crested.readthedocs.io/en/latest/models/BICCN/borzoi_biccn.html).
If you're simply trying to use the architecture without pretrained weights (which is what I suspect you want), you need to import it from [here](https://crested.readthedocs.io/en/latest/api/tools/_autosummary/crested.tl.zoo.borzoi.html), but know that it has many parameters and one of the other architectures will probably work better for simple peak regression.

**Comment by kenxie7** [comment]
Hi! Many thanks for the detailed response, let me add a bit of context here. I wanted to train a model on my own scATAC (multiome) data to generate synthetic enhancers with the peak regression, however since 2114bp peaks are so long for our case that it included a lot of repeat sequences in the region which is not ideal in our application. 

I think point #1 already solves a lot of questions that I had and I think it would already help a lot as a starting point. With the borzoi model, I was trying to follow the tutorial (https://crested.readthedocs.io/en/stable/tutorials/borzoi_atac_finetuning.html) to use the pretrained model on my own data, 1) I initially thought that the peak width could be changed so long as it stays as multiple of 128bp; and secondly 2) when I extracted the downloaded zip file and ran the code, this line causes an error because the h5 file could not be found: "model_weights_path = model_archive.extract("model.weights.h5", tmpdir)" and this was the original issue that I posted. 

Many thanks again!

**Comment by LukasMahieu** [comment]
Ah, now I understand. My apologies, you're right actually. Since Borzoi only uses convolutions and transformer blocks you can retrain with a different input size as long as the final dense layer is retrained (since that one is shape dependent) and it uses multiples of 128bp.
Still, for your point indeed option 1 should work best. I'll take a look why model_archive.extact does not work later, thanks for letting us know.

**Comment by casblaauw** [comment]
Hi, specifically regarding the model weight loading of the resized Borzoi model: the .keras file that you downloaded is actually a zip file which contains the model weights, among other things. [This page](https://keras.io/guides/serialization_and_saving/) has some more info on that. So to access the model weights file, we download the .keras archive (`model_file`) and temporarily extract the model weights file from that. 

Regardless, the real thing you're looking for is the `seq_len` parameter, as Lukas mentioned. This is how we change the model input size for both the CNN models and the bigger Borzoi models.
